### PR TITLE
Ferraille Alexandre - Backend test

### PR DESCRIPTION
**Pull request by alexandre025**
Hi!

This my solution for your backend test, I had fun with it ! 
I've spent around 3 hours reading Twillio's documentation and implementing their ruby sdk. 

I provide a little dashboard with my own gem [cdx](https://rubygems.org/gems/cdx) and I spent extra-time fixing it because the engine wasn't ready for production.

http://aircall.alexandre-ferraille.fr

I'll send you by mail credentials and the phone number.

Thanks!

### CODE REVIEW

File: app/helpers/cdx/admin/calls_helper.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module Cdx
+  module Admin
+    module CallsHelper
+
+      def call_state_tag(call)
```
Comment (gtroppee) [inline app/helpers/cdx/admin/calls_helper.rb]: Could be more concisely written like this

```ruby
def call_state_tag(call)
  color = 
    if %w(failed canceled).include? call.state
      'bg-red'
    elsif %w(ringing in-progress).include? call.state
      'bg-blue'
    elsif call.state == 'completed'
      'bg-green'
    end

  tag.span call.state, class: %w(badge) << color
end
```

File: app/models/call.rb [hunk 2]
```diff
@@ -1,2 +1,29 @@
...
+  validates :state, :from, :to, :call_parameters, presence: true
+
+  # Scopes
+
+  scope :with_state, -> (state) { where(state: state) }
+
+  # Methods
+
+  def duration
+    completed_at ? (completed_at - created_at).round(2) : ''
```
Comment (gtroppee) [inline app/models/call.rb]: Why not set the default duration to `nil` ?

File: app/controllers/calls_controller.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    response.append Twilio::TwiML::Gather.new(num_digits: 1, action: calls_action_url)
+
+    render xml: response
+  end
+
+  # POST /calls/action
+  # Execute to the proper action, selected by a digits
+  def action
+    response = \
+      case params['Digits']
```
Comment (gtroppee) [inline app/controllers/calls_controller.rb]: Best moving this piece of logic in a dedicated model to keep this controller thin.

File: app/controllers/calls_controller.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                         end
+      end
+
+    render xml: response
+  end
+
+  # POST /calls/complete
+  # Mark with a timestamp a completed call
+  def complete
+    @call.update(completed_at: DateTime.now)
```
Comment (gtroppee) [inline app/controllers/calls_controller.rb]: Best using `DateTime.zone.now` or `DateTime.current` -> https://rails-bestpractices.com/posts/2014/10/22/use-time-zone-now-instead-of-time-now/

File: app/helpers/cdx/admin/calls_helper.rb (hunk 1 above)
Comment (alexandre025) [reply app/helpers/cdx/admin/calls_helper.rb]: I didn't know that i could assign something to an if/else statement, thanks

File: app/models/call.rb (hunk 2 above)
Comment (alexandre025) [reply app/models/call.rb]: It's a mistake, set it to nil will make the tests easier

File: app/controllers/calls_controller.rb (hunk 3 above)
Comment (alexandre025) [reply app/controllers/calls_controller.rb]: Indeed, I should use a command pattern like this : https://github.com/andypike/rectify#commands

**Pull request closed without merging**